File: appcenter/ios/Plugin/AppCenterPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(AppCenterPlugin, "AppCenter",
           CAP_PLUGIN_METHOD(getInstallId, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(setUserId, CAPPluginReturnNone);
           CAP_PLUGIN_METHOD(getSdkVersion, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(isEnabled, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(setEnable, CAPPluginReturnNone);
           CAP_PLUGIN_METHOD(setCustomProperties, CAPPluginReturnNone);
           CAP_PLUGIN_METHOD(getLogLevel, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(setLogLevel, CAPPluginReturnNone);
           CAP_PLUGIN_METHOD(setNetworkRequestsAllowed, CAPPluginReturnNone);
           CAP_PLUGIN_METHOD(isNetworkRequestsAllowed, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(setCountryCode, CAPPluginReturnNone);
)

File: appcenter/ios/Plugin/AppCenterPlugin.swift
import Foundation
import Capacitor
import AppCenterCapacitorShared

@objc(AppCenterPlugin)
public class AppCenterPlugin: CAPPlugin {
    private let implementation = AppCenterBase()

    override public func load() {
        AppCenterCapacitorShared.configureWithSettings()
    }

    @objc func getInstallId(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getInstallId()])
    }

    @objc func setUserId(_ call: CAPPluginCall) {
        implementation.setUserId(call.getString("userId") ?? "")
        call.resolve()
    }

    @objc func getSdkVersion(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getSdkVersion()])
    }

    @objc func isEnabled(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.isEnabled()])
    }

    @objc func setEnabled(_ call: CAPPluginCall) {
        implementation.enable(call.getBool("enabled") ?? false)
        call.resolve()
    }

    @objc func getLogLevel(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.getLogLevel()])
    }

    @objc func setLogLevel(_ call: CAPPluginCall) {
        guard let level = call.options["logLevel"] as? Int else {
            call.reject("Must provide LogLevel")
            return
        }

        implementation.setLogLevel(level)
        call.resolve()
    }

    @objc func setCustomProperties(_ call: CAPPluginCall) {
        call.unavailable("Not available in appcenter@2.0.0 or later.")
        //        guard let properties = call.options["properties"] as? [String: [String: Any]] else {
        //            implementation.setCustomProperties([:])
        //            call.resolve()
        //            return
        //        }
        //
        //        implementation.setCustomProperties(properties)
        //
        //        call.resolve()
    }

    @objc func setNetworkRequestsAllowed(_ call: CAPPluginCall) {
        implementation.setNetworkRequestsAllowed(call.getBool("isAllowed", true))
        call.resolve()
    }

    @objc func isNetworkRequestsAllowed(_ call: CAPPluginCall) {
        call.resolve(["value": implementation.isNetWorkRequestsAllowed()])
    }

    @objc func setCountryCode(_ call: CAPPluginCall) {
        implementation.setCountryCode(call.getString("countryCode") ?? "")
        call.resolve()
    }
}
